Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp

```diff
@@ -710,39 +1078,106 @@ DiscusMultipleScatteringCorrection::scatter(const int nScatters, Kernel::PseudoR
   // of a CSGObject sample may not generate valid tracks. Start over again
   // for this event
   if (nlinks == 0) {
-    return {false, 0, 0};
+    return {false, {0.}, 0};
   }
+  std::vector<double> weights;
   const double dl = track.front().distInsideObject;
-  const auto q = (directionToDetector - prevDirection) * kinc;
-  const auto SQ = interpolateFlat(m_SQHist, q.norm());
   if (specialSingleScatterCalc)
     vmu = 0;
   const auto AT2 = exp(-dl * vmu);
   auto scatteringXSectionFull = m_sampleShape->material().totalScatterXSection();
-  weight = weight * AT2 * SQ * scatteringXSectionFull / (4 * M_PI);
-  return {true, weight, QSS};
+  // Step through required overall energy transfer (w) values and work out what
+  // w that means for the final scatter. There will be a single w value for elastic
+  // Slightly different approach to original DISCUS code. It stepped through the w values
+  // in the supplied S(Q,w) distribution and applied each one to the final scatter. If
+  // this resulted in an overall w that equalled one of the required w values it was output.
+  // That approach implicitly assumed S(Q,w)=0 where not specified and that no interpolation
+  // on w would be needed - this may be what's required but seems possible it might not always be
+  for (auto w : wValues) {
+    const double finalE = fromWaveVector(kinc) - w;
+    if (finalE > 0) {
+      const double kout = toWaveVector(finalE);
+      const auto qVector = directionToDetector * kout - prevDirection * k;
+      const double q = qVector.norm();
+      const double finalW = fromWaveVector(k) - finalE;
+      double SQ = Interpolate2D(m_SQWS, finalW, q);
+      weights.emplace_back(weight * AT2 * SQ * scatteringXSectionFull / (4 * M_PI));
+    } else {
+      weights.emplace_back(0.);
+    }
+  }
+  return {true, weights, QSS};
+}
+
+double DiscusMultipleScatteringCorrection::getKf(const double deltaE, const double kinc) {
+  double kf;
+  if (deltaE == 0.) {
+    kf = kinc; // avoid costly sqrt
+  } else {
+    // slightly concerned that rounding errors moving between k and E may mean we take the sqrt of
+    // a negative number in here. deltaE was capped using a threshold calculated using fromWaveVector so
+    // hopefully any rounding will affect fromWaveVector(kinc) in same direction
+    kf = toWaveVector(fromWaveVector(kinc) - deltaE);
```

Comment:
Could we put an assert in to protect from this without incurring a runtime penalty for most users?
That way if it every crashes due to imaginary values a developer running Debug mode will really quickly track it down